cdap-watchdog/src/main/java/io/cdap/cdap/logging/clean/LogCleaner.java

```diff
@@ -71,4 +75,18 @@ public void run() {
     LOG.info("File cleanup completed, Successful file deletes - {}. Failed file deletes - {}. Log Cleanup took {} ms",
              deleteCount, failureCount, (completionTime - startTime));
   }
+
+  private void deleteParentFoldersIfEmpty(String logFilePath) throws IOException {
+    Location logFileLocation = Locations.getLocationFromAbsolutePath(locationFactory, logFilePath);
+    Location folderLocation = Locations.getParent(logFileLocation);
+    while (Locations.LOCATION_COMPARATOR.compare(logsDirectoryLocation, folderLocation) != 0
+            && folderLocation.list().isEmpty()) {
+      if (!folderLocation.delete()) {
```

Comment:
Is there a race condition here? Folder is empty. This thread deletes. A pipeline launching in parallel might assume the folder is there and try to create a log file in it.

Reply:
There is no race condition here.

`Location.delete()` method returns `true` if and only if the file or directory is successfully deleted; `false` otherwise. Handling the case where folder deletion does not succeed due to some reason. Logging it as warning and discontinuing the folder deletion bottom-up.